### DISCUSSION TITLE
Adding extra_attributes to Weather forecast

### DISCUSSION THREAD

**Discussion by kamiyo** [home-assistant/architecture#1080] [Entity Models]
I would like to propose allowing extra attributes back into the weather forecast, this time as a dict on `extra_attributes` attribute. Right now, the weather template limits the attributes forecasts can return, but many services return useful and interesting data (for example, it is possible from NWS to get chance of thunderstorm, wet bulb temperature, etc.).

I am also unsure about the reason for the strict attribute list in the forecasts. Was this because of worries about the size of the state?

There have been some users discussing a bit here: #114799

**Comment by elupus** [comment]
Just add additional entities?

**Reply by MatthewFlamm** [reply]
A forecast might have over 20 values and for an hourly forecast you typically have at a minimum of 24 time points just for hourly forecasts.  We could easily have a thousand entities.

**Reply by joostlek** [reply]
You can use service calls with response values for these kinds of info. We also redirect people to use service responses for things like the energy prices the next day.

**Reply by MatthewFlamm** [reply]
The request is suggesting this, in part.  It is requesting to allow more things to be returned by the existing forecast service call.

If it is 'allowed' to add a custom service for extra forecasts in each integration, integrations will cook up their own flavor leading to inconsistency and issues.

**Reply by joostlek** [reply]
I mean, you can always make a proposal to extend the service. You would then need to support the feature that you want to add with other core integrations that can use it. If none can, should it then be added to the standard?

**Reply by MatthewFlamm** [reply]
The way I'm viewing this is that `extra_attributes` (or whatever it would be) would hold non standard stuff, so there would be a clear demarcation.  But I can appreciate the other view.

Sounds like integrations that want to expose more forecast data should implement their own service, with essentially the same behavior as the built in one just with other attributes.

**Reply by joostlek** [reply]
Right, but then you're still creating a place for non standardized data, but then in standardized data. I would rather see it as a separate service because then people know it's actually specific to this integration and not to expect this at every service.

**Reply by MatthewFlamm** [reply]
It would be ideal from a code owners perspective if the weather integration provides the skeleton for integrations to use, i.e. weather provides the service for the non standard data and integrations provide the data.  But I'm not sure what the exact proposal is yet, so maybe it needs to evolve separately before coalescing if the services overlap significantly.

**Reply by MatthewFlamm** [reply]
It would be good to have some guidance on this topic, even if not an architecture decision, as other integrations are already down the road of the 'thousand entity' approach.  See. https://github.com/home-assistant/core/pull/115847

**Reply by gjohansson-ST** [reply]
Like @joostlek says if NWS has some custom things in its forecasts it should setup its own service to provide it to the users. 

It would be kinda weird to allow suddenly unstructured data in the structure which is in place and nothing stopping integrations to provide this on their own either by entities or by a service with return value.

**Reply by joostlek** [reply]
I also want to mention that I dislike the thousand entity approach. It could also just be a service call with response imo

**Comment by kamiyo** [comment]
I'm not quite getting what you mean. Do you mean to add additional sensors to specific weather integration?

**Comment by elupus** [comment]
Yes

**Comment by kamiyo** [comment]
@MatthewFlamm Do you have any thoughts on this? I really want to try to get those extra bits of forecast in from NWS (from detailed_forecast).

Unless I'm misunderstanding how this would work, for one, that seems to me creating a whole lot of sensors - not to mention to fit in with the forecast model of hourly/daily/twice_daily, you would have to create three separate sensors of each attribute for each type of forecast.

Unless we add a special service to call to get these extra forecast attributes depending on the forecast type. But then I feel like we're just duplicating the `get_forecasts` code, except to get the "unsanctioned" attributes. In which case, it would just be better to use `get_forecasts`. Maybe even putting in the data param for the service call the extra attributes you want to fetch, so it's an opt-in rather than opt-out thing, keeping clutter to a minimum.

Thanks for humoring my thoughts.
